src/main/java/org/candlepin/pki/certs/SCACertificateGenerator.java

```diff
@@ -58,216 +64,342 @@
+
+            X509Certificate x509Cert = createX509Cert(consumer.getUuid(), owner, serial, keypair,
+                container, start, end);
 
-            existing.setSerial(serial);
-            existing.setCert(this.pemEncoder.encodeAsString(x509Cert));
-            this.contentAccessCertificateCurator.saveOrUpdate(existing);
+            scaCertificate.setSerial(serial);
+            scaCertificate.setCert(pemEncoder.encodeAsString(x509Cert));
         }
 
-        Date contentUpdate = owner.getLastContentUpdate();
-        boolean shouldUpdateContent = !contentUpdate.before(existing.getUpdated());
-        if (shouldUpdateContent || isX509CertExpired) {
-            List<Environment> environments = this.environmentCurator.getConsumerEnvironments(consumer);
-            ContentPathBuilder contentPathBuilder = ContentPathBuilder.from(owner, environments);
-            PromotedContent promotedContent = new PromotedContent(contentPathBuilder).withAll(environments);
+        return scaCertificate;
+    }
 
-            Function<ProductContent, String> cidFetcher = pcinfo -> pcinfo.getContent().getId();
+    private ContentAccessPayload getContentAccessPayload(Owner owner, Consumer consumer,
+        List<Environment> environments) throws ConcurrentCertificateCreationException {
 
-            Map<String, ProductContent> ownerContent = this.contentCurator
-                .getActiveContentByOwner(owner.getId())
-                .stream()
-                .collect(Collectors.toMap(cidFetcher, Function.identity(),
-                    (v1, v2) -> new ProductContent(v2.getContent(), v1.isEnabled() || v2.isEnabled())));
+        String payloadKey = new ContentAccessPayloadKeyBuilder(consumer)
+            .setEnvironments(environments)
+            .build();
 
-            byte[] payloadBytes = createContentAccessDataPayload(consumer, ownerContent, promotedContent);
-            existing.setContent(this.createPayloadAndSignature(payloadBytes));
-            this.contentAccessCertificateCurator.saveOrUpdate(existing);
+        ContentAccessPayload container = contentAccessPayloadCurator
+            .getContentAccessPayload(owner.getId(), payloadKey);
+
+        if (container == null || isPayloadExpired(container, owner, environments)) {
+            log.info("Building content access payload for: {}, {}", owner.getKey(), payloadKey);
+
+            Date timestamp = new Date();
+            String payload = buildContentAccessPayload(owner, consumer, environments);
+            if (container == null) {
+                container = new ContentAccessPayload()
+                    .setOwner(owner)
+                    .setPayloadKey(payloadKey)
+                    .setTimestamp(timestamp)
+                    .setPayload(payload);
+
+                try {
+                    contentAccessPayloadCurator.create(container);
+                }
+                catch (EntityExistsException e) {
```

Comment:
Did this work? I was almost certain we'd need to catch an SQLException and then parse the message for "constraint violation" and our constraint name.

Reply:
Also, my fear is that we won't actually get the exception until we commit the transaction -- which means another custom mapper or something. :(

Reply:
I should have functionally tested this, but I didn't and I jumped the gun and assumed the `EntityExistsException` is what I needed to catch, but I think you are right! The SQLException is probably being thrown when we  commit the transaction which is unfortunate. I'll do some testing and adjust this and get some spec tests written!

Reply:
As we discussed online, we won't be able to have spec tests to verify this behavior quite yet, but we'll want them once we get around to updating the specific endpoints which will need to be aware of this behavior.